store/tikv/coprocessor.go

```diff
@@ -677,6 +682,37 @@ func (worker *copIteratorWorker) handleTaskOnce(bo *Backoffer, task *copTask, ch
 	return worker.handleCopResponse(bo, rpcCtx, &copResponse{pbResp: resp.Resp.(*coprocessor.Response)}, task, ch, nil, costTime)
 }
 
+// clientHelper wraps LockResolver and RegionRequestSender.
+// It's introduced to support the new lock resolving pattern in the large transaction.
+// In the large transaction protocol, sending requests and resolving locks are
+// context-dependent. For example, when a send request meets a secondary lock, we'll
+// call ResolveLock, and if the lock belongs to a large transaction, we may retry
+// the request. If there is no context information about the resolved locks, we'll
+// meet the secondary lock again and run into a deadloop.
+type clientHelper struct {
+	*LockResolver
+	*RegionRequestSender
+
+	minCommitTSPushed []uint64
+}
+
+// ResolveLocks wraps the ResolveLocks function and store the resolved result.
+func (ch *clientHelper) ResolveLocks(bo *Backoffer, region RegionVerID, callerStartTS uint64, locks []*Lock) (int64, error) {
+	msBeforeTxnExpired, minCommitTSPushed, err := ch.LockResolver.ResolveLocks(bo, callerStartTS, locks)
+	if err != nil {
+		return msBeforeTxnExpired, err
+	}
+	// Should we deduplicate here?
+	ch.minCommitTSPushed = append(ch.minCommitTSPushed, minCommitTSPushed...)
+	return msBeforeTxnExpired, nil
+}
+
+// SendReqCtx wraps the SendReqCtx function and use the resolved lock result in the kvrpcpb.Context.
+func (ch *clientHelper) SendReqCtx(bo *Backoffer, req *tikvrpc.Request, regionID RegionVerID, sType kv.StoreType) (*tikvrpc.Response, *RPCContext, error) {
+	req.Context.ResolvedLocks = ch.minCommitTSPushed
```

Comment:
from design doc: "The read transaction puts the TS of the uncommitted Lock of the previous has pushed into the context so that it does not need to repeat the Resolve when it encounters the same transaction again."

it seems that we should `minCommitTSPushed` into transaction-level context as the hopeful situation? then all read requests except first one will read previous record without need reslove lock...current impl we only cache pushed ts in cop worker scope..

emmm..it's no easy to cache it in txn scope :confounded: